packages/core/src/messages/challenge.ts

```diff
@@ -0,0 +1,61 @@
+import { u8aEquals, SECP256K1 } from '@hoprnet/hopr-utils'
+import { HASH_ALGORITHM } from './constants'
+import { ecdsaSign, ecdsaVerify, publicKeyCreate } from 'secp256k1'
+import { createHash } from 'crypto'
+
+import type PeerId from 'peer-id'
+
+export class Challenge {
+  private constructor(private ackChallenge: Uint8Array, private signature: Uint8Array) {}
+
+  static get SIZE(): number {
+    return SECP256K1.SIGNATURE_LENGTH
+  }
+
+  static deserialize(preArray: Uint8Array | Buffer, ackChallenge: Uint8Array, pubKey: PeerId): Challenge {
+    if (preArray.length != SECP256K1.SIGNATURE_LENGTH) {
+      throw Error(`Invalid arguments`)
+    }
+
+    let arr: Uint8Array
+    if (typeof Buffer !== 'undefined' && Buffer.isBuffer(preArray)) {
+      arr = Uint8Array.from(preArray)
+    } else {
+      arr = preArray
+    }
+
+    if (!verifyChallenge(pubKey, arr, ackChallenge)) {
+      throw Error(`General error.`)
+    }
+
+    return new Challenge(ackChallenge, arr)
+  }
+
+  serialize(): Uint8Array {
+    return Uint8Array.from(this.signature)
+  }
+
+  static create(ackChallenge: Uint8Array, privKey: PeerId): Challenge {
```

Comment:
if this isn't async, why isn't it in the constructor?

Reply:
It is not implemented as a constructor because it performs a signing operation and produces cryptographic material that is used afterwards to construct the Challenge

Reply:
if it creates a challenge, and has no side effects, it should be the constructor.

Reply:
This does not make any sense because the constructor is used also when deserializing a Challenge, hence the constructor needs two modes, one that signs the challenge and one that takes an array without signing it. This is intentionally impemented as two methods and I don't see any reason to merge both behaviors.